### PR TITLE
[bugfix] Switch to using HMACSHA256 for auth, add legacy cutoff config

### PR DESCRIPTION
**Pull request by CassOnMars**
SHA1 is broken: https://duo.com/decipher/sha-1-fully-and-practically-broken-by-new-collision
also custom HMAC designs can be risky (though this implementation was done well), so this PR switches to use HMACSHA256 and adds a deprecation window

### CODE REVIEW

**Review by sethetter — APPROVED** [review]
This all looks good to me! Tested it locally and it works as intended.

Thanks for putting the legacy cutoff in there. I will say, there's currently one posting live in production, and it was made by @kevinfalting who is also a dev on this project, so if we want to leave the legacy support out to avoid doing the cleanup later, that would be completely fine.

Either way, thank you for this! Definite improvements.

**Comment by kevinfalting** [comment]
Hey, I'm good with not including the legacy support, I don't plan to update the posting. The only bit of feedback I'd give is to use https://pkg.go.dev/crypto/subtle#ConstantTimeCompare instead of rolling our own re: `slowEquals`. ❤️

File: pkg/data/data.go [hunk 1]
```diff
@@ -79,14 +80,13 @@ func (job *Job) Save(db *sqlx.DB) (sql.Result, error) {
...
-		"%s:%s:%s:%s",
+		"%s:%s:%s",
 		job.ID,
 		job.Email,
 		job.PublishedAt.String(),
-		secret,
 	)
 
-	hash := sha1.Sum([]byte(input))
+	hash := hmac.New(sha256.New, []byte(secret)).Sum([]byte(input))
```
Comment (kevinfalting) [inline pkg/data/data.go:89]: Here's a really subtle thing that took me a bit to notice, it has to do with the `.Sum` method on a [`hash.Hash`](https://pkg.go.dev/hash#Hash).

> // Sum appends the current hash to b and returns the resulting slice.
> // It does not change the underlying hash state.
> Sum(b []byte) []byte

The key thing here is that it appends it to b, without any additional hashing. So we can take the resulting hash and just print it out to see the input. Using `Sum` also means the length of the hash is not constant.

Instead, we should use the `Write` method on the hash that is returned from `hmac.New` which would end up looking like:

```go
hash := hmac.New(sha256.New, []byte(secret))
hash.Write([]byte(input))

return base64.URLEncoding.EncodeToString(hash.Sum(nil))
```

**Review by kevinfalting — CHANGES_REQUESTED** [review]

File: pkg/data/data.go (hunk 1 above)
Comment (CassOnMars) [reply pkg/data/data.go:89]: good catch – i always forget how weird golang is about handling hashes.

**Review by kevinfalting — APPROVED** [review]

**Pull request merged**